crates/typst/src/layout/grid/layout.rs

```diff
@@ -677,39 +954,155 @@ impl<'a> GridLayouter<'a> {
+                        // these index operations.
+                        y / 2
+                    })
+                    .map(|hlines| &**hlines)
+                    .unwrap_or(&[]);
+                let tracks = self.rcols.iter().copied().enumerate();
+
+                // Determine all different line segments we have to draw in
+                // this row, and convert them to points and shapes.
+                let segments = generate_line_segments(
+                    self.grid,
+                    tracks,
+                    y,
+                    hlines_at_row,
+                    is_bottom_border,
+                    hline_stroke_at_column,
+                )
+                .into_iter()
+                .map(|(stroke, dx, length)| {
+                    let stroke = (*stroke).clone().unwrap_or_default();
+                    let thickness = stroke.thickness;
+                    let half = thickness / 2.0;
+                    let dx = if self.is_rtl { self.width - dx - length } else { dx };
+                    let target = Point::with_x(length + thickness);
+                    let hline = Geometry::Line(target).stroked(stroke);
+                    (
+                        thickness,
+                        Point::new(dx - half, dy),
                         FrameItem::Shape(hline, self.span),
-                    );
-                }
+                    )
+                });
 
-                // Render vertical lines.
-                for (x, dx) in points(self.rcols.iter().copied()).enumerate() {
-                    let dx = if self.is_rtl { self.width - dx } else { dx };
-                    // We want each vline to span the entire table (start
-                    // at y = 0, end after all rows).
-                    // We use 'split_vline' to split the vline such that it
-                    // is not drawn above colspans.
-                    for (dy, length) in
-                        split_vline(self.grid, rows, x, 0, self.grid.rows.len())
-                    {
-                        let target = Point::with_y(length + thickness);
-                        let vline = Geometry::Line(target).stroked(stroke.clone());
-                        frame.prepend(
-                            Point::new(dx, dy - half),
-                            FrameItem::Shape(vline, self.span),
-                        );
-                    }
-                }
+                // Draw later (after we sort all lines below.)
+                lines.extend(segments);
+            }
+
+            // Sort by increasing thickness, so that we draw larger strokes
+            // on top.
+            // This avoids layering problems where a smaller hline appears
+            // "inside" a larger vline. When both have the same size, hlines
+            // are drawn on top (since the sort is stable, and they are
+            // pushed first).
+            lines.sort_by_key(|(thickness, ..)| *thickness);
```

Comment:
Should we also sort such that user-specified lines are always drawn on top of cell strokes with the same thickness? Right now, user-specified vlines appear below horizontal cell strokes. Note that user-specified vlines have priority over _vertical_ cell strokes, so the argument of having more priority = appearing on top could perhaps not hold here. Or maybe it could.

For now, I kept things as they are, that is, user-specified vlines appearing below horizontal cell strokes with the same thickness (after all, I guess we can also get feedback on this from other table users after the update drops), but wanted to hear other opinions just in case.

Reply:
That is now the case! :)